src/world.rs

```diff
@@ -544,23 +544,37 @@ impl World {
         components: impl DynamicBundle,
     ) -> Result<(), NoSuchEntity> {
         self.flush();
+
         let loc = self.entities.get_mut(entity)?;
+        let old_archetype = loc.archetype;
+        let old_index = loc.index;
 
+        self.insert_bottom_half(entity, components, old_archetype, old_archetype, old_index)
+    }
+
+    fn insert_bottom_half(
+        &mut self,
+        entity: Entity,
+        components: impl DynamicBundle,
+        source_archetype: u32,
```

Comment:
We should carefully document the difference between `source_archetype` and `old_archetype`. Would it be correct to say that the former is the associated archetype graph vertex, whereas the latter is the archetype that the components actually reside in? If so, let's also note that these differ only in the case of fused operations, i.e. `exchange`.

Reply:
Push a proposal for an explanatory doc comment.